sdks/python/apache_beam/transforms/core.py

```diff
@@ -708,6 +708,36 @@ def get_function_arguments(self, func):
 
   def default_type_hints(self):
     fn_type_hints = typehints.decorators.IOTypeHints.from_callable(self.process)
+    batch_fn_type_hints = typehints.decorators.IOTypeHints.from_callable(
+        self.process_batch)
+
+    if fn_type_hints is not None:
```

Comment:
This nested logic is hard to read and a little messy and hard to follow.

For example here and below we have "if fn_type_hints is not None:"

I had to write this a little to really wrap my head around it. I don't know if any parts of my rewrite are useful.

```
# default type hints can come from one of two places. The process type hints or the process batch type hints.
# also the process might be set up to yield batches or to yield elements

has_fn_type_hints = fn_type_hints is not None
has_batch_type_hints = batch_fn_type_hints is not None
yields_batches = self._process_yields_batches
yeilds_elements = self._process_yields_elements

if has_fn_type_hints:
  if yields_elements and has_batch_type_hints:
    # because both batch and element type hints are defined, raise an exception if they are not equal.
    self.validate_batch_and_fn_typehints_same()
    fn_type_hints = fn_type_hints.with_output_types_from(
            batch_fn_type_hints)  # TODO: comment why this is done.
  elif yields_batches:
    # Because the process method produces batches, don't use its output typehints.
    fn_type_hints = fn_type_hints.with_output_types_from(
            typehints.decorators.IOTypeHints.empty())
elif yields_elements and has_batch_type_hints:
  # process_batch produces elements, grab its output typehint
  fn_type_hints = batch_fn_type_hints.with_input_types_from(
            typehints.decorators.IOTypeHints.empty())  # TODO comment why does this use with_input_types_from
```

Reply:
I was a little concerned rewriting this way could break some esoteric use-case, but I ran all our internal tests and it looks good.